PYME/Analysis/points/vector_tools.py

```diff
@@ -0,0 +1,220 @@
+    I = np.argsort(distances, axis=0)
+    # find center of mass of closest n points for each point queried
+    x_com, y_com, z_com = xs[I][:n, :].mean(axis=0), ys[I][:n, :].mean(axis=0), zs[I][:n, :].mean(axis=0)
+    # get directions to the centers of mass
+    azimuth, zenith, r = cartesian_to_spherical(x_com - x_0, y_com - y_0, z_com - z_0)
+
+    v = np.stack([x_com - x_0, y_com - y_0, z_com - z_0])
+    cartesian_vector = v / np.linalg.norm(v, axis=0)
+
+    return azimuth, zenith, r, cartesian_vector
+
+def find_points_within_cone(x, y, z, x0, y0, z0, azimuth, zenith, d_omega=0.15, cutoff_r=1500):
+    # shift and convert to spherical
+    theta, phi, r = cartesian_to_spherical(x - x0, y - y0, z - z0, azimuth, zenith)
+    # tilt points along zenith=0 and look at which are within d_omega
+    inside_cone = abs(phi - zenith) < d_omega
+    inside_cone = np.logical_and(inside_cone, r < cutoff_r)
+    return inside_cone
+
+def find_points_within_cylinder(x, y, z, x0, y0, z0, radius, length, v0, v1, v2):
+    """
+
+    Parameters
+    ----------
+    x
+    y
+    z
+    x0
+    y0
+    z0
+    v0 : ndarray
+        cartesian vector defining the 'axial axis' of the cylinder
+    v1 : ndarray
+        cartesian vector orthogonal to the axial axis
+    v2 : ndarray
+        cartesian vector othogonal to the previous two
+
+    Returns
+    -------
+
+    """
+    # make sure we have a reasonable task
+    assert ((length > 0) and (radius > 0))
+    # make sure all input vectors are unit vectors
+    v_0 = v0 / np.linalg.norm(v0)
+    v_1 = v1 / np.linalg.norm(v1)
+    v_2 = v2 / np.linalg.norm(v2)
+    # make vectors going from the base/center of the cylinder to all of the points we are checking
+    x_c, y_c, z_c = x - x0, y - y0, z - z0
+    positions_c = np.stack([x_c, y_c, z_c]).T
+    # check if they are within the ends of the cylinder, i.e. if the axial component is positive and less than length
+    axial_component = np.dot(positions_c, v_0)
+    inside_axially = np.logical_and(np.greater_equal(axial_component, 0.), np.less_equal(axial_component, length))
+    # now check if they're inside radially
+    radial_component0 = np.dot(positions_c[inside_axially], v_1)
+    radial_component1 = np.dot(positions_c[inside_axially], v_2)
+    # just in case we're working with a ton of points, throw out the ones which are obviously too far out
+    worth_checking = np.logical_and(np.less_equal(radial_component0, radius), np.less_equal(radial_component1, radius))
+    # now do the slow bit
+    inside_r = np.less_equal(np.sqrt(radial_component0[worth_checking] ** 2 + radial_component1[worth_checking] ** 2), radius)
+
+    # handle the indexing
+    inside = np.zeros_like(x, dtype=bool)
+    # can't slice because this copies, use where as a sort of disgusting work around
+    inside[np.where(inside_axially)[0][np.where(worth_checking)[0]][inside_r]] = True
+    return inside, axial_component[inside]
+
+
+def distance_to_image_mask(mask, points):
```

Comment:
This largely duplicates `cluster_morphology.get_labels_from_image()`

Reply:
This largely duplicates `PYME.Analysis.points.objectMeasure.get_labels_from_image()`